File: single-language-analysis/dataset-files/python/test_smell_127.py
# Original URL: https://github.com/ogrisel/scikit-learn/blob/b20998d4ec606558ae86cba7ff5a84aea094c397/sklearn/cluster/tests/test_k_means.py#L83-L111

@pytest.mark.parametrize("array_constr", data_containers, ids=data_containers_ids)
@pytest.mark.parametrize("algo", ["lloyd", "elkan"])
def test_smell_127(array_constr, algo):
    # check that empty clusters are relocated as expected
    X = array_constr([[0, 0], [0.5, 0], [0.5, 1], [1, 1]])

    # second center too far from others points will be empty at first iter
    init_centers = np.array([[0.5, 0.5], [3, 3]])

    kmeans = KMeans(n_clusters=2, n_init=1, init=init_centers, algorithm=algo)
    kmeans.fit(X)

    expected_n_iter = 3
    expected_inertia = 0.25
    assert_allclose(kmeans.inertia_, expected_inertia)
    assert kmeans.n_iter_ == expected_n_iter

    # There are two acceptable ways of relocating clusters in this example, the output
    # depends on how the argpartition strategy breaks ties. We accept both outputs.
    try:
        expected_labels = [0, 0, 1, 1]
        expected_centers = [[0.25, 0], [0.75, 1]]
        assert_array_equal(kmeans.labels_, expected_labels)
        assert_allclose(kmeans.cluster_centers_, expected_centers)
    except AssertionError:
        expected_labels = [1, 1, 0, 0]
        expected_centers = [[0.75, 1.0], [0.25, 0.0]]
        assert_array_equal(kmeans.labels_, expected_labels)
        assert_allclose(kmeans.cluster_centers_, expected_centers)

File: single-language-analysis/dataset-files/java/TestSmell127.java
import org.junit.Test;
import static org.junit.Assert.*;

public class TestSmell127 {

    @Test
    public void testKmeansRelocatedClusters() {
        // check that empty clusters are relocated as expected
        ArrayConstructor arrayConstr = new ArrayConstructor();
        double[][] X = arrayConstr.construct(new double[][]{{0, 0}, {0.5, 0}, {0.5, 1}, {1, 1}});

        // second center too far from others points will be empty at first iter
        double[][] initCenters = new double[][]{{0.5, 0.5}, {3, 3}};

        KMeans kmeans = new KMeans(2, 1, initCenters, "lloyd");
        kmeans.fit(X);

        int expectedNIter = 3;
        double expectedInertia = 0.25;
        assertEquals(expectedInertia, kmeans.getInertia(), 1e-9);
        assertEquals(expectedNIter, kmeans.getNIter());

        // There are two acceptable ways of relocating clusters in this example, the output
        // depends on how the argpartition strategy breaks ties. We accept both outputs.
        try {
            int[] expectedLabels = new int[]{0, 0, 1, 1};
            double[][] expectedCenters = new double[][]{{0.25, 0}, {0.75, 1}};
            assertArrayEquals(expectedLabels, kmeans.getLabels());
            assertArrayEquals(expectedCenters, kmeans.getClusterCenters(), 1e-9);
        } catch (AssertionError e) {
            int[] expectedLabels = new int[]{1, 1, 0, 0};
            double[][] expectedCenters = new double[][]{{0.75, 1.0}, {0.25, 0.0}};
            assertArrayEquals(expectedLabels, kmeans.getLabels());
            assertArrayEquals(expectedCenters, kmeans.getClusterCenters(), 1e-9);
        }
    }
}